thorcast/thorcast.py

```diff
@@ -1,26 +1,79 @@
-#!/usr/bin/env python
-import argparse
-import os
+import time
 
-import yaml
+import redis
+import sqlalchemy
 
-import geocode
-import forecast as fc
+import thorcast.geocode as geocode
+import thorcast.forecast as fc
+import utils.formatters as fmts
 
 
-with open(os.path.join(os.getcwd(), 'config.yml'), 'r') as conffile:
-    config = yaml.safe_load(conffile)
+def lookup(city, state, period, thorcast_conn, redis_conn, logger):
+    """
+    Main API function. Facilitates forecasting from request.
 
-def thorcast(city, state):
-    coordinates = geocode.geocode(city, state, config['GoogleMapsAPIKey'])
-    forecast = fc.get_forecast(**coordinates)
-    return forecast
+    Arguments:
+        city:           [string]:       The city name to forcast
+        state:          [string]:       The state hosting the city
+        period:         [string]:       The day/time to forecast
+        thorcast_conn:  [sqlalchemy.engine.base.Connection]: DB conn
+        redis_conn:     [redis.Redis]:  Redis connection object
+    """
+    period = fmts.sanitize_period(period)
+    city, state = fmts.sanitize_location(city, state)
+    key = f'{city}_{state}_{period}'.lower().replace(' ', '_')
+
+    logger.info(f'Checing Redis for forecast with key {key}')
+    redis_retries = 5
+    while redis_retries:
```

Comment:
probably want to make a retry contextmanager here.
Final use could look like this:
```python
with retries(5, msg='connecting to redis'):
     forecast = redis_conn.lookup(key)
```

Reply:
How can I write a context manager that isn't connected at all to the block it manages?

Reply:
```python
import contextlib
import time

@contextlib.contextmanager
def retry(count):
    while True:
        try:
            yield
            return
        except Exception as e:
            print(f'caught exception: {e}')
            if count == 0:
                raise e
            count -=1

with retry(3):
    print('foo')
    raise 'BOOM!'
```
`RuntimeError: generator didn't stop after throw()`
Python stinks

Reply:
https://trinket.io/python3/0b51abe69e

Reply:
Hm, it also didn't like your renegade exception.

Reply:
https://stackoverflow.com/questions/34775099/why-does-contextmanager-throws-a-runtime-error-generator-didnt-stop-after-thro

Reply:
I saw that before, but didn't want to get into a diatribe about it.
The `with` construct is crippled and could have been so good.
Again, Python stinks.

Reply:
Lol ok, I'll take your word for it. Good thing you won't have to use it at work ;)

Reply:
I'm assuming Ruby allows looping in context managers.

Reply:
Is there a better way I can attempt a persistent connection?

Reply:
Is there a better way I can attempt a persistent connection?